runtime/src/utils/handle.rs

```diff
@@ -198,6 +194,108 @@ impl MetricHandle {
     }
 }
 
+/// A panic that has occurred in a spawned task.
+struct Panic {
+    sender: Option<oneshot::Sender<()>>,
+    panic: Option<Box<dyn Any + Send + 'static>>,
+}
+
+/// Notifies the runtime when a spawned task panics, so it can propagate the failure.
+#[derive(Clone)]
+pub(crate) struct Panicker {
+    catch: bool,
+    inner: Arc<Mutex<Panic>>,
+}
+
+impl Panicker {
+    /// Creates a new [Panicker].
+    pub(crate) fn new(catch: bool) -> (Self, Panicked) {
+        let (sender, receiver) = oneshot::channel();
+        let panic = Arc::new(Mutex::new(Panic {
+            sender: Some(sender),
+            panic: None,
+        }));
+        let panicker = Self {
+            catch,
+            inner: panic.clone(),
+        };
+        let panicked = Panicked {
+            receiver,
+            inner: panic,
+        };
+        (panicker, panicked)
+    }
+
+    /// Returns whether the [Panicker] is configured to catch panics.
+    pub(crate) fn catch(&self) -> bool {
+        self.catch
+    }
+
+    /// Notifies the [Panicker] that a panic has occurred.
+    pub(crate) fn notify(&self, panic: Box<dyn Any + Send + 'static>) {
+        // Log the panic
+        let err = extract_panic_message(&*panic);
+        error!(?err, "task panicked");
+
+        // If we are catching panics, just return
+        if self.catch {
+            return;
+        }
+
+        // If we've already sent a panic, ignore the new one (we check sender
+        // here because we take panic when handling it)
+        let mut inner = self.inner.lock().unwrap();
+        if inner.sender.is_none() {
+            return;
+        }
+
+        // Store the panic and notify `Panicked`
+        inner.panic = Some(panic);
+        let sender = inner.sender.take().unwrap();
+        let _ = sender.send(());
+    }
+}
+
+/// A handle that will be notified when a panic occurs.
+pub(crate) struct Panicked {
+    receiver: oneshot::Receiver<()>,
+    inner: Arc<Mutex<Panic>>,
+}
```

Comment:
Maybe I'm misunderstanding the alerting here, but it seems that `Panicker` will populate the `inner` (which, since it's an `Arc<Mutex<T>>`, also populates it in the `Panicked`, then it sends a unit `()` message over the `oneshot` to tell the `Panicked` to look inside the now-populated `inner`. Is there a reason why it doesn't just send the `Panic` itself over the `oneshot`?

Reply:
Yeah, `oneshot` doesn't guarantee that `receive/next` is ready as soon as `send` returns.

Reply:
Exploring other techniques that feel less jank rn.

Reply:
The "proper" way to probably do this (that oneshot is approximating) is to implement polling directly on `Panicked`. Will make that tweak (will mark "ready for wake" when panic has been set).

Reply:
Upon further research, this approach is actually pretty elegant/minimal. We always need to create some struct to poll over and it more or less just approximates oneshot.

I'm going to just add more comments for now.